### PR TITLE
[WIP] TCPStore listen thread to execute callback function

### PR DESCRIPTION
**Pull request by H-Huang**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#54265 [WIP] TCPStore listen thread to execute callback function**
* #54264 Add watch_key query and listen thread outline

WIP: CI failing in windows, need to use windows alternative to `::poll`

### CODE REVIEW

File: torch/lib/c10d/TCPStore.cpp [hunk 1]
```diff
@@ -204,6 +273,15 @@ void TCPStoreDaemon::compareSetHandler(int socket) {
   } else {
     if (pos->second == currentValue) {
       pos->second = std::move(newValue);
+
```
Comment (pritamdamania87) [inline torch/lib/c10d/TCPStore.cpp:276]: Don't we need to handle `ADD` and `DELETE` as well?

File: torch/lib/c10d/TCPStore.cpp [hunk 2]
```diff
@@ -39,15 +39,69 @@ enum class WatchResponseType : uint8_t { KEY_UPDATED };
...
+  }
 }
 
 void ListenThread::run() {
+  std::vector<struct pollfd> fds;
+  tcputil::addPollfd(fds, controlPipeFd_[0], POLLHUP);
+  tcputil::addPollfd(fds, storeListenSocket_, POLLIN);
+
+  while (1) {
+    SYSCHECK_ERR_RETURN_NEG1(::poll(fds.data(), fds.size(), -1));
```
Comment (pritamdamania87) [inline torch/lib/c10d/TCPStore.cpp:73]: On the server side, doesn't this poll interfere with the poll in TCPStoreDaemon?

**Review by pritamdamania87 — COMMENTED** [review]
We probably need some unit tests for these changes. Also, its probably better to have changes like these in one PR instead of having part of them in https://github.com/pytorch/pytorch/pull/54264.

File: torch/lib/c10d/TCPStore.cpp (hunk 1 above)
Comment (H-Huang) [reply torch/lib/c10d/TCPStore.cpp:288]: Good point, updated the PR

File: torch/lib/c10d/TCPStore.cpp (hunk 2 above)
Comment (H-Huang) [reply torch/lib/c10d/TCPStore.cpp:89]: We now have to create a separate thread on the client to handle responses asynchronously from the master, this is running in that separate thread. I updated the first PR with more details.

**Comment by H-Huang** [comment]
Thanks for reviewing @pritamdamania87, closing this PR and combining the updated changes with #54264 as you mentioned

**Pull request closed without merging**